File: marketplace/openmu/src/GameLogic/Marketplace/EscrowToken.cs
// <copyright file="EscrowToken.cs" company="MUnique">
// Licensed under the MIT License. See LICENSE file in the project root for full license information.
// </copyright>

namespace MUnique.OpenMU.GameLogic.Marketplace;

using System.Buffers.Binary;
using System.Security.Cryptography;
using System.Text;

/// <summary>
/// What the marketplace escrow can be asked to do.
/// </summary>
public enum EscrowOperation : byte
{
    /// <summary>Move an item from the seller's bag into a new escrow box.</summary>
    List = 1,

    /// <summary>Move an item from its escrow box back into the seller's bag.</summary>
    Cancel = 2,

    /// <summary>Take the price from the buyer, move the item into the buyer's bag, leave the proceeds in the box.</summary>
    Buy = 3,

    /// <summary>Move the proceeds of sold boxes into the seller's wallet.</summary>
    Collect = 4,
}

/// <summary>
/// An instruction minted by the marketplace service and signed with the secret it shares
/// with this plugin (MARKETPLACE_ESCROW_SECRET). The client relays it verbatim; the plugin
/// trusts nothing but the signature, the expiry, and the player it is talking to.
/// </summary>
/// <remarks>
/// Layout, little endian:
///   version(1) op(1) expiresAt(8) listingId(16) boxId(16) itemId(16) amount(8) fee(8)
///   slot(1) accountLen(1) account(n) characterLen(1) character(n) hmac(32)
/// </remarks>
public sealed class EscrowToken
{
    private const byte Version = 1;
    private const int MacLength = 32;
    private const int FixedLength = 1 + 1 + 8 + 16 + 16 + 16 + 8 + 8 + 1;

    private EscrowToken()
    {
    }

    public EscrowOperation Operation { get; private set; }

    public DateTimeOffset ExpiresAt { get; private set; }

    public Guid ListingId { get; private set; }

    public Guid BoxId { get; private set; }

    public Guid ItemId { get; private set; }

    /// <summary>The price on Buy and List, the amount on Collect.</summary>
    public long Amount { get; private set; }

    /// <summary>The listing fee on List, the commission on Buy.</summary>
    public long Fee { get; private set; }

    public byte Slot { get; private set; }

    public string Account { get; private set; } = string.Empty;

    public string Character { get; private set; } = string.Empty;

    /// <summary>
    /// Parses and verifies a token. Returns null when the bytes are malformed or the signature is wrong.
    /// The expiry is left to the caller so it can answer with a distinct status.
    /// </summary>
    public static EscrowToken? TryParse(ReadOnlySpan<byte> data, byte[] secret)
    {
        if (data.Length < FixedLength + 2 + MacLength || data[0] != Version)
        {
            return null;
        }

        var body = data[..^MacLength];
        var mac = data[^MacLength..];
        Span<byte> expected = stackalloc byte[MacLength];
        HMACSHA256.HashData(secret, body, expected);
        if (!CryptographicOperations.FixedTimeEquals(mac, expected))
        {
            return null;
        }

        var offset = 1;
        var token = new EscrowToken { Operation = (EscrowOperation)body[offset++] };
        token.ExpiresAt = DateTimeOffset.FromUnixTimeSeconds(BinaryPrimitives.ReadInt64LittleEndian(body.Slice(offset, 8)));
        offset += 8;
        token.ListingId = new Guid(body.Slice(offset, 16));
        offset += 16;
        token.BoxId = new Guid(body.Slice(offset, 16));
        offset += 16;
        token.ItemId = new Guid(body.Slice(offset, 16));
        offset += 16;
        token.Amount = BinaryPrimitives.ReadInt64LittleEndian(body.Slice(offset, 8));
        offset += 8;
        token.Fee = BinaryPrimitives.ReadInt64LittleEndian(body.Slice(offset, 8));
        offset += 8;
        token.Slot = body[offset++];

        var accountLength = body[offset++];
        if (accountLength > 10 || offset + accountLength + 1 > body.Length)
        {
            return null;
        }

        token.Account = Encoding.ASCII.GetString(body.Slice(offset, accountLength));
        offset += accountLength;
        var characterLength = body[offset++];
        if (characterLength > 10 || offset + characterLength != body.Length)
        {
            return null;
        }

        token.Character = Encoding.ASCII.GetString(body.Slice(offset, characterLength));
        return token;
    }
}

File: marketplace/server/escrowToken.ts
import { createHmac, randomUUID } from 'node:crypto';
import {
  EscrowOperation,
  guidToBytes,
  writeInt64LE,
  type EscrowOperationName,
} from '../../src/common/escrowWire';

/**
 * Tokens the game server's escrow plugin acts on. The service mints one per
 * commit (list, cancel, buy, collect); the client relays it untouched; the
 * plugin verifies the signature, the expiry and the player before it moves
 * anything. Layout, little endian, exactly as `EscrowToken.cs` reads it:
 *
 *   version(1) op(1) expiresAt(8) listingId(16) boxId(16) itemId(16)
 *   amount(8) fee(8) slot(1) accountLen(1) account(n) characterLen(1)
 *   character(n) hmac-sha256(32)
 */

export const TOKEN_VERSION = 1;
export const TOKEN_TTL_S = Number(process.env.MARKETPLACE_ESCROW_TOKEN_TTL_S ?? 120);
const MAX_NAME = 10;
const ASCII_NAME_RE = /^[A-Za-z0-9]{1,10}$/;
const NIL_UUID = '00000000-0000-0000-0000-000000000000';

export type EscrowClaims = {
  op: EscrowOperationName;
  listingId: string;
  boxId: string;
  /** Nil for a list, where the item does not have a box yet. */
  itemId?: string;
  /** Price on list and buy; the sum to collect on collect. */
  amount?: number;
  /** Listing fee on list; commission on buy. */
  fee?: number;
  /** Bag slot on list. */
  slot?: number;
  account: string;
  character: string;
  /** Unix seconds; defaults to now plus the TTL. */
  expiresAt?: number;
};

function name(value: string, what: string): Uint8Array {
  if (!ASCII_NAME_RE.test(value)) throw new Error(`escrow token: bad ${what} '${value}'`);
  return new TextEncoder().encode(value);
}

export function mintEscrowToken(claims: EscrowClaims, secret: Uint8Array): Uint8Array {
  const account = name(claims.account, 'account');
  const character = name(claims.character, 'character');
  const expiresAt = claims.expiresAt ?? Math.floor(Date.now() / 1000) + TOKEN_TTL_S;
  const fixed = 1 + 1 + 8 + 16 + 16 + 16 + 8 + 8 + 1;
  const body = new Uint8Array(fixed + 1 + account.length + 1 + character.length);
  let o = 0;
  body[o++] = TOKEN_VERSION;
  body[o++] = EscrowOperation[claims.op];
  writeInt64LE(body, o, expiresAt);
  o += 8;
  body.set(guidToBytes(claims.listingId), o);
  o += 16;
  body.set(guidToBytes(claims.boxId), o);
  o += 16;
  body.set(guidToBytes(claims.itemId ?? NIL_UUID), o);
  o += 16;
  writeInt64LE(body, o, claims.amount ?? 0);
  o += 8;
  writeInt64LE(body, o, claims.fee ?? 0);
  o += 8;
  body[o++] = claims.slot ?? 0;
  body[o++] = account.length;
  body.set(account, o);
  o += account.length;
  body[o++] = character.length;
  body.set(character, o);
  o += character.length;

  const mac = createHmac('sha256', secret).update(body).digest();
  const token = new Uint8Array(body.length + mac.length);
  token.set(body, 0);
  token.set(mac, body.length);
  return token;
}

/** A box id the service mints before the move, so no box can exist unknown to it. */
export function newBoxId(): string {
  return randomUUID();
}

export { MAX_NAME };
